IAM authentication DOCS (#9096)
treeverse/lakeFS: docs/reference/mount.md

Currently, the implemented protocols are `nfs` and `fuse`.
- NFS V3 (Network File System) is supported on macOS.

## Authentication Chain for lakeFS

When running an Everest `mount` command, authentication occurs in the following order:

1. **Session token** from the environment variable `EVEREST_LAKEFS_CREDENTIALS_SESSION_TOKEN` or `LAKECTL_CREDENTIALS_SESSION_TOKEN`.  
   If the token is expired, authentication will fail.
2. **lakeFS key pair**, using lakeFS access key ID and secret key. (picked up from lakectl if Everest not provided)
3. **IAM authentication**, if configured and **no static credentials are set**.

> ## Authentication with lakeFS (using Credentials)

Reviewer: ```suggestion
## Authenticate with lakeFS Credentials
```

The authentication with the target lakeFS server is equal to [lakectl CLI][lakectl].
Searching for lakeFS credentials and server endpoint in the following order:
- Command line flags `--lakectl-access-key-id`, `--lakectl-secret-access-key` and `--lakectl-server-url`
- `LAKECTL_*` Environment variables
- `~/.lakectl.yaml` Configuration file or via `--lakectl-config` flag

> ## Authenticating with lakeFS (using AWS IAM Role)

Reviewer: ```suggestion
## Authenticating with AWS IAM Role
```

- `LAKECTL_*` Environment variables
- `~/.lakectl.yaml` Configuration file or via `--lakectl-config` flag

> ## Authenticating with AWS IAM Role

Reviewer: Link to the IAM feature for people wanting to understand it better (https://docs.lakefs.io/security/external-principals-aws.html)

> ## Authenticating with AWS IAM Role

Reviewer: Link to the IAM feature for people wanting to understand it better (https://docs.lakefs.io/security/external-principals-aws.html)

- `LAKECTL_*` Environment variables
- `~/.lakectl.yaml` Configuration file or via `--lakectl-config` flag

## Authenticating with lakeFS (using AWS IAM Role)

Starting from **lakeFS ≥ v1.57.0** and **Everest ≥ v0.4.0**, authenticating with IAM roles is supported!  

> When IAM authentication is configured, Everest will use your machine's **default AWS profile credentials** to generate a **session token** used for authenticating against lakeFS.  

Reviewer: ```suggestion
When IAM authentication is configured, Everest will use AWS SDK default behavior that will pick your **AWS environment** to generate a **session token** used for authenticating against lakeFS (i.e use `AWS_PROFILE`, `AWS_ACCESS_KEY_ID`, etc) 
```

This token is seamlessly refreshed as long as the AWS session remains valid.  

> For more information regarding IAM authentication, please check [external principals aws](https://docs.lakefs.io/security/external-principals-aws.html).

Reviewer: No need in that last link sentence (I suggested it 1 line below in prerequisites)

> Because IAM roles are linked to users, lakeFS must map each role to a user.  
> To do this, make sure the IAM role is attached to lakeFS. See [Administration of IAM Roles in lakeFS](https://docs.lakefs.io/security/external-principals-aws.html#administration-of-iam-roles-in-lakefs)

Reviewer: That's perfect but add more context: 

```suggestion
### Prerequisites
1. Make sure your lakeFS server supports [AWS IAM Role Login](https://docs.lakefs.io/security/external-principals-aws.html).
2. Make sure your IAM role is attached to lakeFS. See [Administration of IAM Roles in lakeFS](https://docs.lakefs.io/security/external-principals-aws.html#administration-of-iam-roles-in-lakefs)
```

### Authentication Chain

When running an Everest `mount` command, authentication occurs in the following order:

1. **Session token** from the environment variable `EVEREST_LAKEFS_CREDENTIALS_SESSION_TOKEN` or `LAKECTL_CREDENTIALS_SESSION_TOKEN`.  
   If the token is expired, authentication will fail.
2. **lakeFS key pair**, using lakeFS access key ID and secret key. (picked up from lakectl if Everest not provided)
3. **IAM authentication**, if configured.

> ### Configuration

Reviewer: ```suggestion
### Configure everest to use IAM
```

> Since IAM authentication is used as a fallback method, `.lakectl.yaml` **must not contain static credentials**.

Reviewer: IMO you can remove that sentence, above you explain the chain

To use IAM authentication, new configuration fields were introduced:

- `CREDENTIALS.PROVIDER.TYPE = aws_iam`: The identity provider type. Must be equal to `aws_iam`.
- `CREDENTIALS.PROVIDER.AWS_IAM.TOKEN_TTL_SECONDS`: Session token's validation duration.  
- `CREDENTIALS.PROVIDER.AWS_IAM.URL_PRESIGN_TTL_SECONDS`: AWS STS's presigned URL validation duration.  
- `CREDENTIALS.PROVIDER.AWS_IAM.REFRESH_INTERVAL`: The amount of time before token expiration that Everest will try to fetch a new session token instead of using the current one.  

> - `CREDENTIALS.PROVIDER.AWS_IAM.TOKEN_REQUEST_HEADERS`: Map of required `<header>:<value>` to be signed by the AWS STS request.  

Reviewer: ```suggestion
- `CREDENTIALS.PROVIDER.AWS_IAM.TOKEN_REQUEST_HEADERS`: Map of required headers and their values to be signed by the AWS STS request as configured in your lakeFS server. If nothing is set the **default** behavior is adding `x-lakefs-server-id:<lakeFS host>`. If your lakeFS server doesn't require any headers (less secure) you can set this empty by setting `{}` empty map in your config. 

```

> - `CREDENTIALS.PROVIDER.AWS_IAM.TOKEN_REQUEST_HEADERS: x-lakefs-server-id:<lakeFS host>`: Required header.

Reviewer: For all the values - I'm not sure what I'm seeing, `.` and `_` all caps, env vars or viper paths? 
Please stick to the [format we already use](https://docs.lakefs.io/reference/configuration.html) explaining configuration (viper / yaml path) or env var. 

**Format:**
- `key path` `(type: default val)` - [required/optional] extra info  

```
- `credentials.provider.type` `(string: '')` - Settings this `aws_iam` will expect `aws_iam` block and try to use IAM.
- `credentials.provider.aws_iam.token_ttl_second` `(duration: 3600m)` - Optional: lakeFS token duration.
```

> - `CREDENTIALS.PROVIDER.AWS_IAM.TOKEN_REQUEST_HEADERS: x-lakefs-server-id:<lakeFS host>`: Required header.

Reviewer: This should be explanation of above not a standalone thing. x-lakefs-server-id:<lakeFS host>`... 
See comment above for TOKEN_REQUEST_HEADERS

These configuration fields can be set via `.lakectl.yaml`: 

```yaml
credentials:
  provider:
    type: aws_iam          # Required
    aws_iam:
      token_ttl_seconds: 60m              # Optional, default: 1h
      url_presign_ttl_seconds: 15m        # Optional, default: 15m
      refresh_interval: 5m                # Optional, default: 5m

>       token_request_headers:              # Required

Reviewer: ```suggestion
      token_request_headers:              # Optional, if omitted then will set x-lakefs-server-id: <lakeFS host> by default, to override default set to '{}'
```

>         x-lakefs-server-id: <lakeFS host>     # Replace with your actual lakeFS host

Reviewer: Add additional `custom-key: custom-val`  to be clear

>         x-lakefs-server-id: <lakeFS host>     # Replace with your actual lakeFS host

Reviewer: This is confusing since by default it adds this
```suggestion
        # x-lakefs-server-id: <lakeFS host>     Added by default if token_request_headers is not set
```

```

> Or by setting the corresponding environment variables. Those will start with the prefix `EVEREST_LAKEFS_*` or `LAKECTL_*`.

Reviewer: ```suggestion
To set using environment variables - those will start with the prefix `EVEREST_LAKEFS_*` or `LAKECTL_*`.
```

For example, setting the provider type using env vars:
`export EVEREST_LAKEFS_CREDENTIALS_PROVIDER_TYPE=aws_iam` or `LAKECTL_CREDENTIALS_PROVIDER_TYPE=aws_iam`.

> 

Reviewer: Please add a bullet for "tip" for **troubleshooting** presign request adding the presign logging env var specifically to log the presign request

Starting from **lakeFS ≥ v1.57.0** and **Everest ≥ v0.4.0**, authenticating with IAM roles is supported!  
When IAM authentication is configured, Everest will use AWS SDK default behavior that will pick your **AWS environment** to generate a **session token** used for authenticating against lakeFS (i.e use `AWS_PROFILE`, `AWS_ACCESS_KEY_ID`, etc). This token is seamlessly refreshed as long as the AWS session remains valid.  

### Prerequisites
1. Make sure your lakeFS server supports [AWS IAM Role Login](https://docs.lakefs.io/security/external-principals-aws.html).
2. Make sure your IAM role is attached to lakeFS. See [Administration of IAM Roles in lakeFS](https://docs.lakefs.io/security/external-principals-aws.html#administration-of-iam-roles-in-lakefs)

### Authentication Chain

When running an Everest `mount` command, authentication occurs in the following order:

1. **Session token** from the environment variable `EVEREST_LAKEFS_CREDENTIALS_SESSION_TOKEN` or `LAKECTL_CREDENTIALS_SESSION_TOKEN`.  
   If the token is expired, authentication will fail.
2. **lakeFS key pair**, using lakeFS access key ID and secret key. (picked up from lakectl if Everest not provided)

> 3. **IAM authentication**, if configured and **no static credentials are set**.

Reviewer: It's weird that `### Authentication Chain` is inside `## Authenticating with lakeFS`, and thinking about this **whole page** is now a bit confusing -  currently that's the hierarchy and order: 

```md
## Authentication with lakeFS
## Authenticating with lakeFS (using AWS IAM Role)
### Authentication Chain
```

Instead I suggest to do the following order flat hierarchy: 

```md
## Authentication Chain for lakeFS <-- copy from ### Authentication Chain 
## Authentication with lakeFS (using Credentials)
## Authentication with lakeFS (using AWS IAM Role)
```

### Configure everest to use IAM

To use IAM authentication, new configuration fields were introduced:

- `credentials.provider.type` `(string: '')` - Settings this `aws_iam` will expect `aws_iam` block and try to use IAM.
- `credentials.provider.aws_iam.token_ttl_second` `(duration: 60m)` - Optional: lakeFS token duration.
- `credentials.provider.aws_iam.url_presign_ttl_seconds` `(duration: 15m)` - Optional: AWS STS's presigned URL validation duration.  
- `credentials.provider.aws_iam.refresh_interval` `(duration: 15m)` - Optional: Amount of time before token expiration that Everest will try to fetch a new session token instead of using the current one.  
- `credentials.provider.aws_iam.token_request_headers`: Map of required headers and their values to be signed by the AWS STS request as configured in your lakeFS server. If nothing is set the **default** behavior is adding `x-lakefs-server-id:<lakeFS host>`. If your lakeFS server doesn't require any headers (less secure) you can set this empty by setting `{}` empty map in your config. 

These configuration fields can be set via `.lakectl.yaml`: 

> ```yaml

Reviewer: For completness of the example add the lakeFS host to the YAML as well
